src/evqueue_api.8
.TH "EVQUEUE_API" 8 "2016-07-11" "evQueue utilities" "evqueue"

.SH NAME
evqueue_api \- evQueue utilities
.SH "SYNOPSIS"
.PP
\fBevqueue_api\fR [ OPTIONS ] --<group> --<action> [--<parameter name> <parameter value>]*
.SH "DESCRIPTION"
.PP
Utility to query evQueue API from command line. Can be used to connect to the evQueue core through TCP or UNIX sockets.
.PP
Options, group and action MUST be passed before other arguments. Arguments expected after action are expected to be API command parameters only.
.SH "OPTIONS"
.TP
\fB--connect\fR \fIconnection_string\fR
Connect to server identified by \fIconnection_string\fR. Two formats are accepted :

tcp socket: tcp://<hostname>[:<port>]

unix socket: unix://<path>
.TP
\fB--user <username>\fR
Use this username if authentication is required by daemon.
.TP
\fB--password <password>\fR
Use this password if authentication is required by daemon.
.TP
\fB--noformat\fR
Returns raw XML instead of indented XML.
.TP
\fB--xpath\fR
Returns the first result of the XPath expression specified.
.TP
\fB--<group>\fR
The API command group, for example --tasks to query task API.
.TP
\fB--<action>\fR
The action to request to API command group, for example --get to query an element.
.SH "CONFIGURATION FILES"
.PP
evqueue_api will read configuration files in /etc/evqueue.api.conf and ~/.evqueue.api.conf
.PP
The following options are recognized (but will be overridden by command line parameters) : api.connect, api.user, api.password, api.noformat
.SH EXAMPLES
evqueue_api --task --get --id 1
.br
Retrieve task information from API.

evqueue_api --connect unix:///var/run/evqueue.socket --user admin --password admin --workflows --list
.br
Connect to evqueue engine using UNIX socket and authentication. Request the list of available workflows.
.SH SEE ALSO
.BR http://www.evqueue.net/



debian/changelog
evqueue-core (3.3-1) buster; urgency=medium

  * ELogs
  * Bugfixes

 -- Thibault KUMMER <bob@coldsource.net>  Wed, 18 May 2022 13:50:00 +0200

evqueue-core (3.2-1) buster; urgency=medium

  * Automatic tags
  * Bugfixes

 -- Thibault KUMMER <bob@coldsource.net>  Mon, 14 Dec 2020 10:54:16 +0200

evqueue-core (3.1-1) buster; urgency=medium

  * Bugfixes

 -- Thibault KUMMER <bob@coldsource.net>  Tue, 10 Nov 2020 12:30:16 +0200

evqueue-core (3.0-1) buster; urgency=medium

  * Websockets support

 -- Thibault KUMMER <bob@coldsource.net>  Thu, 14 May 2020 12:30:16 +0200

evqueue-core (2.1-1) buster; urgency=medium

  * Fix: run subjobs when loop has no iterations
  * Fix: segfault
  * Fix: restart rsyslog only if present (closes: #881596)
  * Added debconf translations (closes: #882654, #883290, #883302, #883822,
    #884171)

 -- Thibault KUMMER <bob@coldsource.net>  Thu, 08 Nov 2018 09:52:40 +0000

evqueue-core (2.0-2) buster; urgency=medium

  * Initial release (closes: #880973).

 -- Thibault KUMMER <bob@coldsource.net>  Wed, 17 Aug 2016 12:30:16 +0200


src/evqueue_api.cpp
/*
 * This file is part of evQueue
 * 
 * evQueue is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * evQueue is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with evQueue. If not, see <http://www.gnu.org/licenses/>.
 * 
 * Author: Thibault Kummer <bob@coldsource.net>
 */

#include <stdio.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>

#include <API/ClientBase.h>
#include <Exception/Exception.h>
#include <XML/XMLFormatter.h>
#include <XML/XMLString.h>
#include <DOM/DOMDocument.h>
#include <Configuration/ConfigurationReader.h>
#include <Configuration/Configuration.h>

#include <map>
#include <string>
#include <algorithm>
#include <memory>
#include <iostream>
#include <sstream>

using namespace std;

static void usage()
{
	fprintf(stderr,"Usage : evqueue_api --<group> --<action> [--<parameter name> <parameter value>]*\n");
	fprintf(stderr,"  --connect <cnx string>\n");
	fprintf(stderr,"  --user <username>\n");
	fprintf(stderr,"  --password <password>\n");
	fprintf(stderr,"  --noformat\n");
	fprintf(stderr,"  --xpath\n");
	exit(-1);
}

int main(int argc, char  **argv)
{
	int exit_status = 0;
	
	try
	{
		// Default config
		Configuration config({{"api.connect","tcp://localhost:5000"}, {"api.user", ""}, {"api.password", ""}, {"api.noformat", "no"}, {"api.xpath", ""}});
		
		// Try to read from default paths
		ConfigurationReader::ReadDefaultPaths("evqueue.api.conf", &config);
		
		// Override with command line
		int cur = ConfigurationReader::ReadCommandLine(argc, argv, {"connect", "user", "password", "bool:noformat", "xpath"}, "api", &config);
		if(cur==-1)
			usage();

		string connection_str = config.Get("api.connect");
		string user = config.Get("api.user");
		string password = config.Get("api.password");
		bool format = !config.GetBool("api.noformat");
		string xpath = config.Get("api.xpath");

		// If we received a password, compute sha1 of it as a real password
		if(password.length())
			password = ClientBase::HashPassword(password);

		// Read group and action
		if(cur+1>=argc)
			usage();

		if(strncmp(argv[cur],"--",2)!=0)
			usage();
		const char *group = argv[cur]+2;

		if(strncmp(argv[cur+1],"--",2)!=0)
			usage();
		const char *action = argv[cur+1]+2;

		cur+=2;

		// Read parameters
		map<string,string> parameters;
		for(;cur<argc;cur+=2)
		{
			if(strncmp(argv[cur],"--",2)!=0)
				usage();
			
			string name = argv[cur];
			string value;
			
			if(cur+1>=argc)
			{
				// Try to read last argrument from stdin
				ostringstream ss;
				ss<<cin.rdbuf();
				value = ss.str();
			}
			else
				value = argv[cur+1];
			
			parameters[name] = value;
		}

		// Build API Query XML
		xercesc::XMLPlatformUtils::Initialize();

		{
			DOMDocument xmldoc;
			DOMElement root_node = xmldoc.createElement(group);
			xmldoc.appendChild(root_node);
			
			root_node.setAttribute("action",action);
			for(auto it=parameters.begin();it!=parameters.end();it++)
			{
				string name = it->first;
				name = name.substr(2);
				replace( name.begin(), name.end(), '-', '_');
				root_node.setAttribute(name,it->second);
			}
				
			string query_xml = xmldoc.Serialize(xmldoc.getDocumentElement());
			
			// Send API command to evQueue
			{
				ClientBase client(connection_str,user,password);
				client.Exec(query_xml);
				
				DOMDocument *xmldoc = client.GetResponseDOM();
				
				if(xpath!="")
				{
					unique_ptr<DOMXPathResult> res(xmldoc->evaluate(xpath, xmldoc->getDocumentElement(), DOMXPathResult::FIRST_RESULT_TYPE));
					if(res->length()>0)
						printf("%s\n", res->getStringValue().c_str());
				}
				else
				{
					string response_xml = xmldoc->Serialize(xmldoc->getDocumentElement());
					
					if(format)
					{
						XMLFormatter formatter(response_xml);
						formatter.Format();
					}
					else
						printf("%s\n",response_xml.c_str());
				}
			}
		}
	}
	catch(Exception &e)
	{
		fprintf(stderr,"%s",e.error.c_str());
		if(e.code!="")
			fprintf(stderr," (%s)",e.code.c_str());
		fprintf(stderr,"\n");
		exit_status = -1;
	}
	
	xercesc::XMLPlatformUtils::Terminate();
	
	return exit_status;
}
